src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+            }
+
+            // A negated discard node represents an always false pattern
+            private bool IsNegatedDiscard(BoundPattern pattern, bool synthesized = false)
+            {
+                if (synthesized)
+                {
+                    return pattern is BoundNegatedPattern { Negated: BoundDiscardPattern { WasCompilerGenerated: true } };
+                }
+                else
+                {
+                    return pattern is BoundNegatedPattern { Negated: BoundDiscardPattern };
+                }
+            }
+
+            public override BoundNode? VisitBinaryPattern(BoundBinaryPattern node)
+            {
+                var resultLeft = (BoundPattern)Visit(node.Left);
+                var resultRight = (BoundPattern)Visit(node.Right);
+                var result = node.WithLeft(resultLeft).WithRight(resultRight);
+                return _negated ? result.WithDisjunction(!node.Disjunction) : result;
+            }
+
+            public override BoundNode? VisitNegatedPattern(BoundNegatedPattern node)
+            {
+                if (_negated)
+                {
+                    return Instance.Visit(node.Negated);
+                }
+
+                return NegatedInstance.Visit(node.Negated);
+            }
+
+            public BoundPattern NegateIfNeeded(BoundPattern node)
+            {
+                if (_negated)
+                {
+                    var result = new BoundNegatedPattern(node.Syntax, node, node.InputType, narrowedType: node.InputType);
+                    if (node.WasCompilerGenerated)
+                    {
+                        result.MakeCompilerGenerated();
+                    }
+                    return result;
+                }
+
+                return node;
+            }
+
+            public override BoundNode? VisitTypePattern(BoundTypePattern node)
+            {
+                return NegateIfNeeded(node);
+            }
+
+            public override BoundNode? VisitConstantPattern(BoundConstantPattern node)
+            {
+                return NegateIfNeeded(node);
+            }
+
+            public override BoundNode? VisitDiscardPattern(BoundDiscardPattern node)
+            {
+                return NegateIfNeeded(node);
+            }
+
+            public override BoundNode? VisitDeclarationPattern(BoundDeclarationPattern node)
+            {
+                BoundPattern result;
+                if (node.IsVar)
+                {
+                    result = MakeDiscardPattern(node).MakeCompilerGenerated();
+                }
+                else
+                {
+                    if (node.InputType.Equals(node.DeclaredType.Type)) // TODO2 what kind of type comparison?
+                    {
+                        result = MakeDiscardPattern(node).MakeCompilerGenerated();
+                    }
+                    else
+                    {
+                        result = node;
```

Comment:
>result = node; [](http://example.com/codeflow?start=24&length=14)

I see that this line is the source of a warning for the following scenario. However, I think that the warning is undesirable in this case. It feels like the warning would be desirable only in a situation when the declaration pattern cannot be satisfied. It looks like we usually get an "error CS8121" for scenarios like that in `is` expression, perhaps there are non-error scenarios in `switch` expression. 
```
        [Fact]
        public void Test_03()
        {
            var source = """
var c = new C();

if (c is C and {F: object x}) // 1
{
    x.ToString();
}

class C
{
#pragma warning disable CS0649
    public int F;
}
""";
            var comp = CreateCompilation(source);
            comp.VerifyEmitDiagnostics(
                // (3,17): warning CS9269: The pattern is redundant.
                // if (c is C and {F: object x}) // 1
                Diagnostic(ErrorCode.WRN_RedundantPattern, "F: object x").WithLocation(3, 17)
                );
        }
```